File: src/gossip_glomers/broadcast.clj
(ns gossip-glomers.broadcast
  (:require [gossip-glomers.node :as node]))

;; Index every broadcast message we receive. Each node sends messages to other
;; nodes along with their index, so that they know if they've missed any. Each
;; node also periodically tells every other node the last index it saw, so that
;; it can be sent any missed messages.
;;
;; This strategy never re-sends messages unnecessarily, and for parts 3d and 3e,
;; results in:
;;
;;   messages per operation: ~13
;;   median latency: ~70ms
;;   maximum latency: ~100ms

(defn init [state]
  (swap! state assoc :messages (zipmap (:node-ids @state) (repeat [])))
  (future
    (loop []
      (Thread/sleep 10000)
      (run! #(node/post state % {:type "sync"
                                 :index (count ((:messages @state) %))})
            (remove #{(:node-id @state)} (:node-ids @state)))
      (recur))))

(defn handle [state msg]
  (case (:type (:body msg))
    "broadcast"
    (let [index (count ((:messages @state) (:node-id @state)))]
      (swap! state update-in [:messages (:node-id @state)]
             conj (:message (:body msg)))
      (run! #(node/post state % {:type "new_messages"
                                 :messages [(:message (:body msg))]
                                 :index index})
            (remove #{(:node-id @state)} (:node-ids @state)))
      (node/reply state msg {:type "broadcast_ok"}))

    "read"
    (node/reply state msg
                {:type "read_ok"
                 :messages (distinct (apply concat (vals (:messages @state))))})

    "topology"
    (node/reply state msg {:type "topology_ok"})

    "new_messages"
    (when (= (count ((:messages @state) (:src msg))) (:index (:body msg)))
      (swap! state update-in [:messages (:src msg)]
             into (:messages (:body msg))))

    "sync"
    (let [messages ((:messages @state) (:node-id @state))]
      (when (< (:index (:body msg)) (count messages))
        (node/post state (:src msg)
                   {:type "new_messages"
                    :messages (subvec messages (:index (:body msg)))
                    :index (:index (:body msg))})))))

(defn -main []
  (node/run init handle))

File: src/broadcast.js
#! /usr/bin/env -S deno run

// Results for parts d and e:
//
//   - messages per operation: ~19
//   - median latency: ~300ms
//   - maximum latency: ~500ms

import { Node } from "./node.js";
import { randomElement } from "./utils.js";

const node = new Node(new Set());

node.run({
  init: (_) => {
    if (node.peers.length == 0) {
      return;
    }
    setTimeout(
      () =>
        setInterval(() => {
          node.send(randomElement(node.peers), {
            type: "gossip",
            state: [...node.state],
          });
        }, 20),
      100
    );
  },
  topology: (msg) => node.reply(msg, { type: "topology_ok" }),
  broadcast: (msg) => {
    node.state.add(msg.body.message);
    node.reply(msg, { type: "broadcast_ok" });
  },
  read: (msg) =>
    node.reply(msg, {
      type: "read_ok",
      messages: [...node.state],
    }),
  gossip: (msg) => {
    for (const message of msg.body.state) {
      node.state.add(message);
    }
  },
});
